File: src/portfolio_manager/repositories/account_repository.py
"""Account repository for database operations."""

from datetime import datetime, timezone
from decimal import Decimal
from typing import Any
from uuid import UUID, uuid4

from portfolio_manager.models import Account
from portfolio_manager.services.database import AccountModel

_UNSET: Any = object()


class AccountRepository:
    """Repository for Account database operations."""

    def create(self, name: str, cash_balance: Decimal) -> Account:
        """Create a new account."""
        now = datetime.now(timezone.utc)
        row = AccountModel.create(
            id=uuid4(),
            name=name,
            cash_balance=cash_balance,
            created_at=now,
            updated_at=now,
        )
        return self._to_domain(row)

    def list_all(self) -> list[Account]:
        """List all accounts."""
        return [self._to_domain(row) for row in AccountModel.select()]

    def delete_with_holdings(self, account_id: UUID, holding_repository) -> None:
        """Delete an account and its holdings."""
        holding_repository.delete_by_account(account_id)
        AccountModel.delete().where(AccountModel.id == account_id).execute()

    def update(
        self,
        account_id: UUID,
        name: str,
        cash_balance: Decimal,
        kis_account_no: str | None = _UNSET,
        kis_api_key_id: int | None = _UNSET,
    ) -> Account:
        """Update an account name and cash balance."""
        now = datetime.now(timezone.utc)
        fields: dict = {"name": name, "cash_balance": cash_balance, "updated_at": now}
        if kis_account_no is not _UNSET:
            fields["kis_account_no"] = kis_account_no
        if kis_api_key_id is not _UNSET:
            fields["kis_api_key_id"] = kis_api_key_id
        AccountModel.update(**fields).where(AccountModel.id == account_id).execute()

        row = AccountModel.get_by_id(account_id)
        return self._to_domain(row)

    @staticmethod
    def _to_domain(row: AccountModel) -> Account:
        return Account(
            id=row.id,
            name=row.name,
            cash_balance=Decimal(str(row.cash_balance)),
            created_at=row.created_at,
            updated_at=row.updated_at,
            kis_account_no=row.kis_account_no,
            kis_api_key_id=row.kis_api_key_id,
        )

File: internal/repositories/account_repository.go
package repositories

import (
	"context"
	"database/sql"
	"errors"

	"github.com/kadragon/portfolio-manager/internal/db/sqlc"
	"github.com/kadragon/portfolio-manager/internal/ktime"
	"github.com/kadragon/portfolio-manager/internal/models"
	"github.com/kadragon/portfolio-manager/internal/numeric"
	"github.com/kadragon/portfolio-manager/internal/uuidx"
)

// AccountRepository provides account CRUD operations.
type AccountRepository struct {
	q *sqlc.Queries
}

// NewAccountRepository builds an AccountRepository over the given queries handle.
func NewAccountRepository(q *sqlc.Queries) *AccountRepository {
	return &AccountRepository{q: q}
}

// Create inserts a new account with zero cash balance by default.
func (r *AccountRepository) Create(ctx context.Context, name string, cashBalance numeric.Decimal) (models.Account, error) {
	now := ktime.Now()
	row, err := r.q.CreateAccount(ctx, sqlc.CreateAccountParams{
		ID:          uuidx.New(),
		Name:        name,
		CashBalance: cashBalance,
		CreatedAt:   now,
		UpdatedAt:   now,
	})
	if err != nil {
		return models.Account{}, err
	}
	return toAccount(row), nil
}

// ListAll returns all accounts in insertion order.
func (r *AccountRepository) ListAll(ctx context.Context) ([]models.Account, error) {
	rows, err := r.q.ListAccounts(ctx)
	if err != nil {
		return nil, err
	}
	out := make([]models.Account, 0, len(rows))
	for _, row := range rows {
		out = append(out, toAccount(row))
	}
	return out, nil
}

// GetByID returns the account with the given id, or nil if absent.
func (r *AccountRepository) GetByID(ctx context.Context, id uuidx.UUID) (*models.Account, error) {
	row, err := r.q.GetAccountByID(ctx, id)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	a := toAccount(row)
	return &a, nil
}

// UpdateNameCash updates only name and cash_balance (used by bulk-cash; preserves KIS fields).
func (r *AccountRepository) UpdateNameCash(ctx context.Context, id uuidx.UUID, name string, cashBalance numeric.Decimal) (models.Account, error) {
	row, err := r.q.UpdateAccountNameCash(ctx, sqlc.UpdateAccountNameCashParams{
		Name:        name,
		CashBalance: cashBalance,
		UpdatedAt:   ktime.Now(),
		ID:          id,
	})
	if err != nil {
		return models.Account{}, err
	}
	return toAccount(row), nil
}

// Update sets all editable fields including broker bindings and account type.
func (r *AccountRepository) Update(
	ctx context.Context,
	id uuidx.UUID,
	name string,
	cashBalance numeric.Decimal,
	kisAccountNo sql.NullString,
	kisAPIKeyID sql.NullInt64,
	accountType sql.NullString,
	tossAccountSeq sql.NullInt64,
) (models.Account, error) {
	row, err := r.q.UpdateAccount(ctx, sqlc.UpdateAccountParams{
		Name:           name,
		CashBalance:    cashBalance,
		KisAccountNo:   kisAccountNo,
		KisApiKeyID:    kisAPIKeyID,
		AccountType:    accountType,
		TossAccountSeq: tossAccountSeq,
		UpdatedAt:      ktime.Now(),
		ID:             id,
	})
	if err != nil {
		return models.Account{}, err
	}
	return toAccount(row), nil
}

// DeleteWithHoldings removes holdings then the account (two sequential statements, parity with Python).
func (r *AccountRepository) DeleteWithHoldings(ctx context.Context, id uuidx.UUID) error {
	if err := r.q.DeleteHoldingsByAccount(ctx, id); err != nil {
		return err
	}
	return r.q.DeleteAccount(ctx, id)
}

func toAccount(row sqlc.Account) models.Account {
	var kisAccountNo *string
	if row.KisAccountNo.Valid {
		kisAccountNo = &row.KisAccountNo.String
	}
	var kisAPIKeyID *int64
	if row.KisApiKeyID.Valid {
		kisAPIKeyID = &row.KisApiKeyID.Int64
	}
	var accountType *string
	if row.AccountType.Valid {
		accountType = &row.AccountType.String
	}
	var tossAccountSeq *int64
	if row.TossAccountSeq.Valid {
		tossAccountSeq = &row.TossAccountSeq.Int64
	}
	return models.Account{
		ID:             row.ID,
		Name:           row.Name,
		CashBalance:    row.CashBalance,
		CreatedAt:      row.CreatedAt.Time,
		UpdatedAt:      row.UpdatedAt.Time,
		KisAccountNo:   kisAccountNo,
		KisAPIKeyID:    kisAPIKeyID,
		AccountType:    accountType,
		TossAccountSeq: tossAccountSeq,
	}
}
